torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -0,0 +1,293 @@
+  if (THPVariable_Check(value)) {
+    return reinterpret_cast<THPVariable*>(value)->cdata;
+  }
+  if (THPUtils_checkLong(value)) {
+    return type.scalarTensor(Scalar(THPUtils_unpackLong(value)));
+  }
+  if (PyFloat_Check(value)) {
+    return type.scalarTensor(Scalar(THPUtils_unpackDouble(value)));
+  }
+  throw TypeError("can't assign a %s to a %s", Py_TYPE(value)->tp_name, type.toString());
+}
+
+static Variable applySlicing(const Variable& self, PyObject* index, variable_list& outIndices) {
+  int64_t size = PyTuple_GET_SIZE(index);
+  int64_t dim = 0;
+
+  auto handle_var = [&](const Variable& var) {
+    // TODO: check scalarType
+    outIndices.resize(dim + 1);
+    outIndices[dim] = var;
+    dim++;
+  };
+
+  Variable result = self;
+  for (int64_t i = 0; i < size; i++) {
+    PyObject* obj = PyTuple_GET_ITEM(index, i);
+    if (THPUtils_checkLong(obj)) {
+      result = applySelect(result, dim, THPUtils_unpackLong(obj));
+    } else if (PySlice_Check(obj)) {
+      result = applySlice(result, dim, obj);
+      dim++;
+    } else if (obj == Py_Ellipsis) {
+      int64_t unspecified_dims = self.dim() - count_specified_dimensions(index);
+      for(; unspecified_dims > 0; unspecified_dims--) {
+        dim++;
+      }
+    } else if (obj == Py_None) {
+      result = result.unsqueeze(dim);
+      dim++;
+    } else if (THPVariable_Check(obj)) {
+      handle_var(reinterpret_cast<THPVariable*>(obj)->cdata);
+    } else if (THPModule_isTensor(obj)) {
+      handle_var(make_variable(createTensor(obj)));
+    } else if (PySequence_Check(obj)) {
+      handle_var(sequenceToVariable(self.type(), obj));
+    } else {
+      invalid_index(obj);
+    }
+  }
+  return result;
+}
+
+static std::vector<Tensor> asTensorList(const variable_list& v) {
+  return std::vector<Tensor>(v.begin(), v.end());
+}
+
+static Variable dispatch_index(const Variable& self, const variable_list& indices) {
+  AutoNoGIL no_gil;
+  AutoGPU auto_gpu(self);
+  return self.index(asTensorList(indices));
+}
+
+static Variable dispatch_index_put_(Variable& self, const variable_list& indices, const Variable& value) {
+  AutoNoGIL no_gil;
+  AutoGPU auto_gpu(self);
+  return self.index_put_(asTensorList(indices), value);
+}
+
+static bool treatSequenceAsTuple(PyObject* index) {
+  if (PyTuple_Check(index)) {
+    return true;
+  }
+  if (!PySequence_Check(index)) {
+    return false;
+  }
+  // This uses a heuristics from NumPy for determining whether to treat
+  // non-tuple sequences as if they were a tuple. From the NumPy code comments:
+  //
+  // "At this point, we're left with a non-tuple, non-array, sequence:
+  //  typically, a list. We use some somewhat-arbitrary heuristics from here
+  //  onwards to decided whether to treat that list as a single index, or a
+  //  list of indices. Backwards compatibility only takes effect for short
+  //  sequences - otherwise we treat it like any other scalar."
+  auto n = PySequence_Size(index);
```

Comment:
When would this be negative?

Reply:
Okay, can we add that as a comment. In general having negative return values for these errors seems like it happens in Python, but its a little mentally incongruous when the operation itself is getting what we would expect to be a non-negative number.